sklearn/cross_validation.py

```diff
@@ -1042,13 +1042,21 @@ def cross_val_predict(estimator, X, y=None, cv=None, n_jobs=1,
                                                       train, test, verbose,
                                                       fit_params)
                             for train, test in cv)
-    p = np.concatenate([p for p, _ in preds_blocks])
+
+    p = [p for p, _ in preds_blocks]
     locs = np.concatenate([loc for _, loc in preds_blocks])
     if not _check_is_partition(locs, _num_samples(X)):
         raise ValueError('cross_val_predict only works for partitions')
-    preds = p.copy()
-    preds[locs] = p
-    return preds
+
+    # Check for sparse predictions
+    if sp.issparse(p[0]) :
+        preds = sp.vstack(p, format = 'csr')
+        return preds
+    else :
+        p = np.concatenate(p)
+        preds = p.copy()
+        preds[locs] = p
```

Comment:
Sorry for my naivety, I now realize that it would be needed in the sparse case as well.

Although I encounter an efficiency issue while doing the same in the sparse case. 
On running the following code snippet I get the following warning.
```python
    a = sp.csr_matrix([[0,0],[0,1],[1,1]])
    id = numpy.array([2,0,1])
    b = a.copy()
    b[id] = a
```
    /Library/Python/2.7/site-packages/scipy/sparse/compressed.py:739: SparseEfficiencyWarning: Changing the sparsity structure of a csr_matrix is expensive. lil_matrix is more efficient.
  SparseEfficiencyWarning)

Also on printing b I get the following 
    print b
  (0, 0)	0
  (0, 1)	1
  (1, 0)	1
  (1, 1)	1
  (2, 0)	0
  (2, 1)	0
  
which is obviously inefficient.
@jnothman 
Can you please suggest any other efficient way of achieving the same or point out if I am going wrong somewhere.